notebooks/stitches_dev/functions/gridded_recipe.R

```diff
@@ -0,0 +1,203 @@
+    
+  }) %>% 
+    dplyr::bind_rows() -> 
+    out 
+  
+  return(out)
+  
+}
+
+
+
+# Go through a recipe and ensure that all of the periods have the same archive 
+# and target period length, if not update to reflect the target period length. 
+# Otherwise you'll end up with extra years in the stitched data. This is really 
+# only an issue for the final period of target data because sometimes that period is somewhat short. 
+# Args 
+# rp: a data frame of the recipe 
+# Return: a recipe data frame that has target and archive periods of the same length.
+handle_final_period <- function(rp){
+  
+  out <- data.frame(stringsAsFactors = FALSE)
+  
+  # For each row in rp check to see if the length 
+  # of the target and the archive years are the same. 
+  for(row in 1:nrow(rp)){
+    
+    len_target <- rp$target_end_yr[row] - rp$target_start_yr[row]
+    len_archive <- rp$archive_end_yr[row] - rp$archive_start_yr[row]
+    
+    if(len_target == len_archive){
+      
+      out <- rbind(out, rp[row, ])
+      
+    } else if (len_target < len_archive) {
+      
+      # Figure out how much shorter the target period is than the archive period. 
+      diff <- len_archive - len_target
+      updated <- rp[row, ]
+      updated$archive_end_yr <- updated$archive_end_yr - 1
+      
+      # Add the updated row to the data frame. 
+      out <- rbind(out, updated)
+    } else {
+      
+      stop("don't know how to handle this problem!")
+      
+    }
+  }
+  
+  return(out)
+  
+}
+
+
+
+
+# Generate recipe for the stitched gridded product 
+# (based on the notebook 5 code)
+# Args 
+#   target_data: a data frame of the target data 
+#   archive_data: the data frame of the archive data to match on
+#   n: int number of permutations based on the neighborhood matching TODO write some test to account for when tol is small 
+#   tol: default set to 0.1, it defines the threshold for the neighborhood matching process
+# Return: data frame containing the following columns, stitching_id, target_start_yr, target_end_yr, 
+#     archive_start_yr, archive_end_yr, and file. This is all of the information that will be needed 
+#     by python to stitch the gridded product. 
+generate_gridded_recipe <- function(target_data, archive_data, n, tol = 0.1){
+  
+  # Load the complete archive. 
+  complete_archive <- read.csv("inputs/main_raw_pasted_tgav_anomaly_all_pangeo_list_models.csv", 
+                               stringsAsFactors = FALSE) %>% 
+    dplyr::select(-X) %>% 
+    mutate(experiment = ifelse(grepl('historical', file), 'historical', experiment)) %>%  
+    select(model, experiment, ensemble, variable, file) %>%  
+    distinct()
+  
+  # This code is taken from Notebook 5
+  # ----------------------------------------------------------------
+  # Get all your matches in your neighborhood and convert them to
+  # the number of recipes you want:
+  match_neighborhood(target_data = target_data,
+                     archive_data = archive_subset,
```

Comment:
should be `archive_data=archive_data`; it works right now because `archive_subset` is globally defined before calling `generate_gridded_recipes`